### PR TITLE
[GraphBolt] Check return values of liburing calls and refactor.

### PR DESCRIPTION
**Pull request by mfbalin**
## Description

In the old code, we were not checking the return value of the results. If the reads are short (requested N bytes but it reads M < N bytes), then we have to continue from where it left off and submit another request. Now, we can write a regression test 
and measure speed and after speed is measured, check correctness without affecting timing.

Regression test:

```python
import os
import tempfile
from functools import partial
import time

import torch
import dgl.graphbolt as gb

assert_equal = partial(torch.testing.assert_close, rtol=0, atol=0)

def to_on_disk_numpy(test_dir, name, t):
    path = os.path.join(test_dir, name + ".npy")
    gb.numpy_save_aligned(path, t.numpy())
    return path

def test_index_select_throughput_and_iops(shape, dtype, indices, num_threads_list):
    tensor = torch.randint(0, 127, shape, dtype=dtype)

    skip_first = 10

    results = []
    IOPSs = []

    with tempfile.TemporaryDirectory() as test_dir:
        path = to_on_disk_numpy(test_dir, "tensor", tensor)

        for num_threads in num_threads_list:
            feature = gb.DiskBasedFeature(path=path, num_threads=num_threads)

            throughput_sum = 0
            iops_sum = 0

            for i, idx in enumerate(indices):
                start = time.time()
                result = feature.read(idx)
                duration = time.time() - start
                assert_equal(result, tensor[idx])

                if i >= skip_first:
                    throughput_sum += result.nbytes / duration
                    iops_sum += idx.numel() / duration
    
            throughput = throughput_sum / (len(indices) - skip_first)
            iops = iops_sum / (len(indices) - skip_first)
            print(num_threads, int(throughput / (2 ** 20)), "MiB/s", int(iops), "IOPS")
            results.append(throughput)
            IOPSs.append(iops)

    return results, IOPSs

shape = [2500000, 4096]
dtype = torch.int8
batch_size = 100000
indices = [torch.randint(0, shape[0], [batch_size], dtype=torch.int32) for _ in range(25)]
num_threads_list = list(range(1, 9))

throughputs, IOPSs = test_index_select_throughput_and_iops(shape, dtype, indices, num_threads_list)
print(list((num_threads, int(throughput / (2 ** 20)), int(iops)) for num_threads, throughput, iops in zip(num_threads_list, throughputs, IOPSs)))
```

Benchmark results with 4K byte feature dimension:
```
(venv) mfbalin@BALIN-PC:~/dgl-1$ python graphbolt/benchmarks/disk_based_feature.py
1 715 MiB/s 183271 IOPS
2 1142 MiB/s 292576 IOPS
3 1571 MiB/s 402287 IOPS
4 1785 MiB/s 457213 IOPS
5 1873 MiB/s 479614 IOPS
6 2094 MiB/s 536154 IOPS
7 2185 MiB/s 559578 IOPS
8 2228 MiB/s 570410 IOPS
[(1, 715, 183271), (2, 1142, 292576), (3, 1571, 402287), (4, 1785, 457213), (5, 1873, 479614), (6, 2094, 536154), (7, 2185, 559578), (8, 2228, 570410)]
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mfbalin** [comment]
@dgl-bot

File: graphbolt/src/cnumpy.h [hunk 1]
```diff
@@ -5,50 +5,34 @@
...
 #include <fstream>
 #include <iostream>
 #include <memory>
 #include <string>
+#include <vector>
 
 namespace graphbolt {
 namespace storage {
 
-template <typename T>
```
Comment (mfbalin) [inline graphbolt/src/cnumpy.h:29]: @frozenbugs see deleted code here. This needs to be deleted and io_uring code requires extra changes.

File: graphbolt/src/cnumpy.cc [hunk 2]
```diff
@@ -124,6 +120,29 @@ c10::intrusive_ptr<Future<torch::Tensor>> OnDiskNpyArray::IndexSelect(
 #endif  // HAVE_LIBRARY_LIBURING
 }
 
+struct ReadRequest {
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:123]: nit, struct refers to data struct with no method, otherwise class.

File: graphbolt/src/cnumpy.cc [hunk 3]
```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
...
     const auto thread_id = begin;
-    // Record the inside offsets of fetched features.
-    int64_t residual[kGroupSize];
-    while (true) {
-      begin = work_queue.fetch_add(kGroupSize, std::memory_order_relaxed);
-      if (begin >= index.numel() || error_flag.load()) break;
-      end = std::min(begin + kGroupSize, index.numel());
+    auto &my_io_uring_queue = io_uring_queue_[thread_id];
+    auto my_read_buffer = ReadBuffer(thread_id);
+    // The completion queue might contain 4 * kGroupSize while we may submit
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:166]: why 4?

File: graphbolt/src/cnumpy.cc [hunk 4]
```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
...
+    // 4 * kGroupSize more. No harm in overallocation here.
+    CircularQueue<ReadRequest> read_queue(8 * kGroupSize);
+    int64_t num_submitted = 0;
+    int64_t num_completed = 0;
+    auto Submit = [&](int64_t submission_minimum_batch_size) {
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:171]: or submit_fn
```suggestion
    auto submit = [&](int64_t submission_minimum_batch_size) {
```

File: graphbolt/src/cnumpy.cc [hunk 5]
```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
...
+      while (!read_queue.IsEmpty()) {
+        const auto submitted = ::io_uring_submit(&my_io_uring_queue);
+        TORCH_CHECK(submitted >= 0);
+        num_submitted += submitted;
+        // Pop the submitted entries from the queue.
+        read_queue.PopN(submitted);
+      }
+    };
+    for (int64_t read_buffer_slot = 0; true;) {
+      auto RequestReadBuffer = [&]() {
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:187]: ditto.
```suggestion
      auto request_read_buffer = [&]() {
```

File: graphbolt/src/cnumpy.cc (hunk 3 above)
Comment (mfbalin) [reply graphbolt/src/cnumpy.cc:166]: See the comment in line 70-71. It says
```c++
    // We have allocated 2 * kGroupSize submission queue entries and
    // 4 * kGroupSize completion queue entries after this call.
```

File: graphbolt/src/cnumpy.cc [hunk 6]
```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
...
   torch::parallel_for(0, num_thread_, 1, [&](int64_t begin, int64_t end) {
+    if (begin >= end) return;
     const auto thread_id = begin;
-    // Record the inside offsets of fetched features.
-    int64_t residual[kGroupSize];
-    while (true) {
-      begin = work_queue.fetch_add(kGroupSize, std::memory_order_relaxed);
-      if (begin >= index.numel() || error_flag.load()) break;
-      end = std::min(begin + kGroupSize, index.numel());
+    auto &my_io_uring_queue = io_uring_queue_[thread_id];
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:165]: my? Why it is not yours :)

File: graphbolt/src/cnumpy.cc [hunk 7]
```diff
@@ -137,86 +156,141 @@ torch::Tensor OnDiskNpyArray::IndexSelectIOUringImpl(torch::Tensor index) {
...
+      // Submit and wait for the reads.
+      while (!read_queue.IsEmpty()) {
+        const auto submitted = ::io_uring_submit(&my_io_uring_queue);
+        TORCH_CHECK(submitted >= 0);
+        num_submitted += submitted;
+        // Pop the submitted entries from the queue.
+        read_queue.PopN(submitted);
+      }
+    };
+    for (int64_t read_buffer_slot = 0; true;) {
```
Comment (frozenbugs) [inline graphbolt/src/cnumpy.cc:187]: nit: The whole IndexSelectIOUringImpl has nested many layers, it is better to to breakout to sub-function for better readability.

**Review by frozenbugs — APPROVED** [review]

File: graphbolt/src/cnumpy.cc (hunk 6 above)
Comment (mfbalin) [reply graphbolt/src/cnumpy.cc:165]: The class one is `io_uring_queue_`. I thought I could call it `io_uring_queue` but it is too similar to the class member. When thinking of a name, I impersonated the threads we create in the constructor and thought this queue belongs to the thread. Thus, `my_` prefix :).

File: graphbolt/src/cnumpy.cc (hunk 7 above)
Comment (mfbalin) [reply graphbolt/src/cnumpy.cc:187]: Let me put this to my todo list.

**Pull request merged**